---
title: 'KCNI Summer 2024 Workshop: Part 1'
author: "Derek"
date: "`r Sys.Date()`"
output: github_document
---

# Schedule
Part 1: Intro to single-cell RNAseq analysis, R, and Seurat

Part 2: Intro to differential expression, cell type identification & visualizations

Part 3: Case-control differential expression with pseudobulks

# Datasets
https://drive.google.com/drive/folders/1VOGGuPjDT49vz95mk4huWOIX9s9-FIQ8?usp=sharing
Put the datasets in the `data/` directory of this project.

# Resources
## Basic R:
R for Data Science
https://r4ds.hadley.nz/

## Seurat:
Intro tutorial: 
https://satijalab.org/seurat/articles/pbmc3k_tutorial.html

Differential expression testing:
https://satijalab.org/seurat/articles/de_vignette.html

Data visualization:
https://satijalab.org/seurat/articles/visualization_vignette.html

Dataset integration & related analyses:
https://satijalab.org/seurat/articles/integration_introduction.html

Cell type annotation:
https://satijalab.org/seurat/articles/integration_mapping.html

## Case-control differential expression with pseudobulks:
https://hbctraining.github.io/scRNA-seq/lessons/pseudobulk_DESeq2_scrnaseq.html

# Intro to single-cell RNAseq analysis, R, and Seurat
## Setup

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(here)
library(tidyverse)
library(Seurat)
```

### Load data 
Let's start by loading in the metadata and gene counts for the human Smart-seq dataset.
Note that loading the counts matrix might take a couple minutes to run.

```{r load-data}
# metadata
meta_smart <- read.csv(here("data", "AIBS_smart_meta_mini.csv"), row.names = 1)

# counts matrix
counts_smart <- read.csv(here("data/AIBS_smart_counts_mini.csv"), row.names = 1) 
```

## Intro to tidyverse
magrittr: pipe *%>%*

dplyr:
- mutate
- filter
- select
- group_by
- summarize
- pull

**And later, we can look at...**

dplyr:
- bind_rows, left_join, etc.

tidyr:
- pivot_longer/pivot_wider

ggplot2:
- plot types (e.g. histogram, boxplot, bar, smooth, point, violin)
- themes and aesthetics

### Pipe %>% 
These three operations are equivalent:
```{r pipe-example}
# how many columns are in our metadata?
ncol(meta_smart) 

# instead of directly listing our metadata object within the function call, "pipe" 
# it in to the desired location, marked by "."
meta_smart %>%  
  ncol(.)

# when piping into a function, the target location is assumed to be the first 
# parameter, so we don't need to specify "." unless we want it somewhere else 
# (or in multiple locations)
meta_smart %>% 
  ncol()
```

### dplyr 
#### mutate()
Let's say we want to add a column to our metadata that stores both the subclass label and the cortical layer information for each cell:
```{r mutate-example}
meta_smart <- meta_smart %>% 
  mutate(subclass_layer_label = paste(subclass_label, cortical_layer_label, 
                                      sep = " "))
```

#### filter()
Okay, now we're only interested in keeping cells annotated as VIP neurons:
```{r filter-example}
vip_neurons <- meta_smart %>% 
  filter(subclass_label == "VIP")
```
How many VIP cells are we left with?

#### select()
We don't need all of these columnsâ€”let's only keep the ones we've just used, plus the sample name. Also, columns will appear in the order you list them, so this is an easy opportunity to reorder.

```{r select-example}
selected_data <- meta_smart %>% 
  select(sample_name, subclass_label, cortical_layer_label)
```

#### group_by()
Let's say we want to apply a function to info from our metadata in a group-wise manner. For example, we want to count up the number of cells that we have within each cortical layer.

```{r group-by-example}
cell_counts <- meta_smart %>% 
  group_by(cortical_layer_label) %>% 
  tally() # tallies the number of items in each unique group 
```

#### summarize()
This function works similarly to mutate in that it will create new columns that you define, but instead of creating a new value for each row like mutate, it will collapse to one row containing a value that summarizes all rows or, when provided with a grouping structure, a set of rows each containing a value that summarizes all rows within a group. 

Let's put this all together to ask a really specific (albeit odd) question with our metadata:
What region / cortical layer combo is the most common location for our VIP neurons, stratified by donor sex?
```{r summarize-example}
vip_summary <- meta_smart %>% 
  filter(subclass_label == "VIP") %>% 
  select(region_label, cortical_layer_label, donor_sex_label) %>% 
  mutate(region_layer_label = paste(region_label, cortical_layer_label, 
                                    sep = " ")) %>% 
  group_by(donor_sex_label, region_layer_label) %>% 
  tally() %>% 
  summarize(region_summary = region_layer_label[which.max(n)])
```

#### pull
Finally, if there's a column in a dataframe that you'd like to grab and use somewhere else, you can extract those values using this function.

For example, you might just want to grab the sample names to use for filtering other dataframes:
```{r pull-example}
sample_names <- meta_smart %>% 
  pull(sample_name)
```

## Intro to the Seurat object
We'll start off with the human SMART-seq data we loaded earlier.

For more information about this dataset, check here: https://portal.brain-map.org/atlases-and-data/rnaseq/human-multiple-cortical-areas-smart-seq

```{r create-seurat-object}
Seu_smart <- CreateSeuratObject(counts = t(counts_smart), 
                                meta.data = meta_smart) 
# t(counts_smart) transposes so that we have rows as genes and columns as samples as they should be for Seurat

# can choose to remove large matrices now that we have Seurat obj:
#rm(counts_smart, meta_smart) 
```

*S3 vs. S4 objects*
- Seurat obj is S4, S4s much more defined, must contain and can only contain certain things which are formally named
- informal dataframes and lists like metadata stored in S3

We can move through data structure with *$* or *@* (can use tab completion to know which is appropriate)

```{r seurat-object-exploration}
Idents(Seu_smart) # interacts with active.ident part of seurat obj. By default 
                  # without providing orig.ident, pulls string before _
Seu_smart@active.ident # same as above
```

### Pre-processing
We're first going to go through each step one-by-one to understand what is happening and what we might want to change for our dataset, and then we'll check out an all-in-one method afterwards.

Note: normally, it is bad practice to overwrite an object when making changes to it. Here, we are simply adding results in data section of the Seurat object, not overwriting any data, because we are using Seurat functions.

#### Normalization
```{r normalization}
Seu_smart <- NormalizeData(Seu_smart, normalization.method = "LogNormalize", 
                           scale.factor = 1000000) # changing scale.factor to mil 
                                                   # so we get cpm

# look at normalized data
Seu_smart@assays$RNA$data[100:105, 100:105]

```

#### Find variable features
```{r find-variable-features}
Seu_smart <- FindVariableFeatures(Seu_smart, #selection.method = "vst", 
                                  nfeatures = 2000) # should see effect of 
                                                    # changing nfeatures

# look at most variable features
head(VariableFeatures(Seu_smart))
```

#### Scale data
```{r scale-data}
Seu_smart <- ScaleData(Seu_smart, verbose = FALSE)
```

#### Run Principal Component Analysis (PCA)
```{r run-pca}
Seu_smart <- RunPCA(Seu_smart, npcs = 50, verbose = FALSE) # 50 is default, we could 
# choose something smaller based on ElbowPlot below

ElbowPlot(Seu_smart, ndims = 50) # see SD of each PC, shows how much variance explained 
# use to see how many PC needed to best explain data
# cut at the elbow (can argue where cutoff is, might choose 7 or 20)
```

#### Find neighbors
```{r find-neighbors}
Seu_smart <- FindNeighbors(Seu_smart, reduction = "pca", dims = 1:20) 
# default dims is 1:10
```

#### Find clusters
```{r find-clusters}
Seu_smart <- FindClusters(Seu_smart, resolution = 0.5) # default resolution is 0.8
# nm.method and annoy.metric have drastic effects on cluster creation

# tells you number of cells in each cluster
table(Seu_smart$seurat_clusters) 

# number of cells per class per cluster
table(Seu_smart$seurat_clusters, Seu_smart$class_label) 

# number of cells per subclass per cluster
table(Seu_smart$seurat_clusters, Seu_smart$subclass_label) 
```

#### Run UMAP
```{r run-umap}
Seu_smart <- RunUMAP(Seu_smart, reduction = "pca", dims = 1:20)

# visualizing clusters
p1 <- DimPlot(Seu_smart, reduction = "umap", group.by = "subclass_label", label = TRUE)
p2 <- DimPlot(Seu_smart, reduction = "umap", group.by = "seurat_clusters", label = TRUE, 
              repel = TRUE)

p1 + p2 # to view side-by-side
```

# Save your Seurat object for next time
```{r save-seurat-object}
dir.create(here("data/processed"), showWarnings = FALSE, recursive = TRUE)
saveRDS(Seu_smart, here("data/processed/Seu_smart.rds"))
```