---
  title: 'KCNI Summer 2024 Workshop: Part 2'
author: "Derek"
date: "`r Sys.Date()`"
output: github_document
---

### Load packages  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(here)
library(tidyverse)
library(Seurat)
library(ggpubr)
```
  
### Load Seurat object
```{r}
Seu_smart <- readRDS(here("data/processed/Seu_smart.rds"))
```

## Finding cluster marker genes
### Set your active identity
This changes the grouping variable for your Seurat object for all of the following analyses. Because we are going to conduct comparisons between clusters, let's use the new column that FindClusters() added to our metadata called "seurat_clusters".

```{r}
Idents(Seu_smart) <- "seurat_clusters" #set active identity to our newly defined clusters
```

### Perform differential expression
*FindMarkers()* identifies genes that are differentially expressed between one cluster and all other clusters. The parameters ident.1, logfc.threshold, and min.pct specify the cluster to compare against, the minimum log-fold change threshold, and the minimum percentage of cells expressing the gene, respectively. This step helps identify genes that can potentially distinguish one cluster from other clusters, and can be used for cell type annotation.

We can use this same function to find genes that can be used to:
- Distinguish one cluster from another closely related cluster
- Distinguish one cluster from all other clusters present

*Note the defaults for this function:* logfc.threshold = 0.25, min.pct = 0.1

*Consider:* can also choose to set a value for min.diff.pct (which is -Inf by default). This allows you to only test genes with a minimum difference in the min.pct values between the compared groups.

```{r}
# get genes to distinguish cluster 8 from 3
cluster_8_v_3 <- FindMarkers(Seu_smart, ident.1 = 8, ident.2 = 3, 
                             logfc.threshold = log(2), min.pct = 0.5)

# genes to distinguish cluster 8, 2, or 13 from all other clusters
cluster_8_v_all <- FindMarkers(Seu_smart, ident.1 = 8, logfc.threshold = log(2), 
                               min.pct = 0.5)
cluster_2_v_all <- FindMarkers(Seu_smart, ident.1 = 2, logfc.threshold = log(2), 
                               min.pct = 0.5)
cluster_13_v_all <- FindMarkers(Seu_smart, ident.1 = 13, logfc.threshold = log(2), 
                                min.pct = 0.5)
```

*FindAllMarkers()* will find markers that distinguish each cluster from the others for all clusters at once. If you're looking to identify cell types for your clusters, this will be the one you want to run. Note that it takes a while.

```{r}
all_clusters <- FindAllMarkers(Seu_smart, logfc.threshold = log(2), min.pct = 0.50)
```

### Examine results
```{r}
# taking a look at most significantly DE marker genes
cluster_8_v_3 %>% head(n = 10) %>% knitr::kable()
cluster_8_v_all %>% head(n = 10) %>% knitr::kable()
cluster_2_v_all %>% head(n = 10) %>% knitr::kable()

# two different ways of looking for a specific gene in the output dataframes, 
# where genes are rownames
cluster_2_v_all %>% 
  filter(row.names(.)=="GRIN1")

cluster_2_v_all %>% 
  rownames_to_column("gene_name") %>% 
  filter(gene_name=="GRIN1")
```

## Using metadata for other comparisons
Now, assume that we've performed cell type annotation for all of our cells using the clusters we've defined and their marker genes, and these are now stored in "subclass_label" in our metadata.

Let's say we aren't just interested in comparing between cell types, but also comparing between cell types derived from male vs. female donors. Could also think to do something like this for a cell type-specific case-control comparison.

Let's focus in on cluster 8, which we would've identified as microglia. 
```{r}
# making new metadata column to see if any sex differences for the same cell type
Seu_smart@meta.data <- Seu_smart@meta.data %>% 
  mutate(sex_subclass = paste(donor_sex_label, subclass_label, sep="_"))

table(Seu_smart$sex_subclass) #see how many cells we have for each sex+subclass combo

Idents(Seu_smart) <- "sex_subclass" #setting this new column as our active identity

unique(Idents(Seu_smart)) #seeing what our options are for making comparisons

#finding genes that are DE in female-derived microglia vs male-derived
F_microglia_vs_M_microglia <- FindMarkers(Seu_smart, ident.1 = "F_Microglia", 
                                          ident.2 = "M_Microglia", 
                                          logfc.threshold = log(2), min.pct = 0.25)

F_microglia_vs_M_microglia %>% head(n = 10) %>% knitr::kable()

Seu_smart@assays$RNA$counts["GFAP",][1:10] %>% knitr::kable() #checking if a gene is present in matrix
#this is grabbing the number of reads per sample for the gene GFAP for samples 1-10
```

## Visualizations
We'll continue with cluster 8 (microglia) to generate a few different figures with Seurat's plotting functions.
```{r}
#getting top 6 marker genes for distinguishing cluster 8 cells, saving to plot below
features <- cluster_8_v_all %>% 
  head(n=6) %>% 
  row.names

Idents(Seu_smart) <- "seurat_clusters" #setting our active identity back to our clusters
```

### Violin Plot
*VlnPlot()* generates a violin plot to visualize the expression distribution of the selected marker genes (features) across different clusters. This plot provides insights into the expression patterns and variations of these genes within each cluster, including the specific expression levels and the presence of bimodal or skewed distributions.

```{r}
VlnPlot(Seu_smart, features = features)
```

### Feature Plot
*FeaturePlot()* creates a feature plot to visualize the expression of the selected marker genes (features) in a scatterplot-like format. This plot displays the expression levels of the genes on the x-axis and the fraction of cells expressing the genes on the y-axis, with each dot representing a cell. It helps visualize the cell-to-cell variation in gene expression and identify potential correlations or differences between clusters.

```{r}
FeaturePlot(Seu_smart, features = features)
```

### Dot Plot
*DotPlot()* generates a dot plot to visualize the expression of the selected marker genes (features) in a matrix-like format. Each row represents a gene, each column represents a cluster, and the size/color of the dots indicates the gene expression level. This plot helps identify genes that are specifically expressed or enriched in cluster 8 compared to other clusters.

```{r}
DotPlot(Seu_smart, features = features) + RotatedAxis()
```

### Heat Map
*DoHeatmap()* creates a heat map to visualize the expression patterns of the selected marker genes (features) across cells. The heat map represents the gene expression values as a color-coded matrix, with rows corresponding to genes and columns corresponding to cells. It provides a comprehensive view of gene expression similarities and differences across cells and clusters.

```{r}
DoHeatmap(subset(Seu_smart, downsample = 100), features = features, size = 3, slot="data") +
  scale_fill_viridis_c()
```
